File: adapters/typescript/src/rfc8949.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { encodeRfc8949, EncodeError } from "./rfc8949.ts";
import { hexEncode } from "./util.ts";
import type { LogicalValue } from "./logicalValue.ts";

function hex(v: LogicalValue): string {
  return hexEncode(encodeRfc8949(v));
}

function Int(value: string): LogicalValue {
  return { type: "int", value };
}
function Text(value: string): LogicalValue {
  return { type: "text", value };
}
function Bytes(value: string): LogicalValue {
  return { type: "bytes", value };
}
function Bool(value: boolean): LogicalValue {
  return { type: "bool", value };
}
const Null: LogicalValue = { type: "null" };
function Arr(items: LogicalValue[]): LogicalValue {
  return { type: "array", items };
}
function Map_(entries: [LogicalValue, LogicalValue][]): LogicalValue {
  return { type: "map", entries };
}
function Float(width: "auto" | "f16" | "f32" | "f64", value: string): LogicalValue {
  return { type: "float", width, value };
}
function Tag(tag: bigint, value: LogicalValue): LogicalValue {
  return { type: "tag", tag, value };
}
function Bignum(sign: "positive" | "negative", value: string): LogicalValue {
  return { type: "bignum", sign, value };
}

test("shortestIntZero", () => {
  assert.equal(hex(Int("0")), "00");
});

test("shortestIntBoundary23_24", () => {
  assert.equal(hex(Int("23")), "17");
  assert.equal(hex(Int("24")), "1818");
});

test("shortestIntBoundary255_256", () => {
  assert.equal(hex(Int("255")), "18ff");
  assert.equal(hex(Int("256")), "190100");
});

test("shortestIntBoundary65535_65536", () => {
  assert.equal(hex(Int("65535")), "19ffff");
  assert.equal(hex(Int("65536")), "1a00010000");
});

test("shortestIntBoundary4294967295_4294967296", () => {
  assert.equal(hex(Int("4294967295")), "1affffffff");
  assert.equal(hex(Int("4294967296")), "1b0000000100000000");
});

test("negativeInt", () => {
  assert.equal(hex(Int("-1")), "20");
});

test("negativeIntMinI64Boundary", () => {
  assert.equal(hex(Int("-9223372036854775808")), "3b7fffffffffffffff");
});

test("textAndBytes", () => {
  assert.deepEqual(encodeRfc8949(Text("a")), new Uint8Array([0x61, 0x61]));
  assert.equal(hex(Bytes("deadbeef")), "44deadbeef");
});

test("boolAndNull", () => {
  assert.deepEqual(encodeRfc8949(Bool(true)), new Uint8Array([0xf5]));
  assert.deepEqual(encodeRfc8949(Bool(false)), new Uint8Array([0xf4]));
  assert.deepEqual(encodeRfc8949(Null), new Uint8Array([0xf6]));
});

test("arrayOfInts", () => {
  assert.deepEqual(encodeRfc8949(Arr([Int("1"), Int("2")])), new Uint8Array([0x82, 0x01, 0x02]));
});

test("nanCanonicalPayload", () => {
  assert.equal(hex(Float("auto", "NaN")), "f97e00");
});

test("nanCanonicalPayloadUnderExplicitWidthForcing", () => {
  for (const width of ["f16", "f32", "f64"] as const) {
    assert.equal(hex(Float(width, "NaN")), "f97e00");
  }
});

test("negativeZeroPreservedDistinctFromZero", () => {
  const neg = hex(Float("auto", "-0.0"));
  const pos = hex(Float("auto", "0.0"));
  assert.equal(neg, "f98000");
  assert.equal(pos, "f90000");
  assert.notEqual(neg, pos);
});

test("shortestFloatFormF16Exact", () => {
  assert.equal(hex(Float("auto", "2.5")), "f94100");
});

test("explicitFloatWidthForcesEncoding", () => {
  assert.equal(hex(Float("f64", "2.5")), "fb4004000000000000");
});

test("explicitFloatWidthRaisesOnPrecisionLoss", () => {
  assert.throws(() => encodeRfc8949(Float("f16", "0.1")), EncodeError);
});

test("mapKeysSortedByEncodedBytes", () => {
  assert.equal(hex(Map_([[Int("9"), Int("1")], [Int("10"), Int("2")]])), "a209010a02");
});

test("mapKeysPureBytewiseNoLengthPresort", () => {
  // -24 encodes to 1 byte (0x37); 1000 encodes to 3 bytes (0x1903e8). Pure bytewise
  // order (no length pre-sort) means 0x19 < 0x37, so the LONGER encoding of 1000
  // sorts FIRST. A length-first (RFC 7049-style) comparator would wrongly sort
  // -24 first. This is the critical regression trap for this task.
  assert.equal(
    hex(Map_([[Int("-24"), Int("1")], [Int("1000"), Int("2")]])),
    "a21903e8023701",
  );
});

test("bignumAt2Pow64UsesTag2", () => {
  assert.equal(hex(Bignum("positive", "18446744073709551616")), "c249010000000000000000");
});

test("negativeBignumOffsetByOne", () => {
  assert.equal(hex(Bignum("negative", "18446744073709551616")), "c348ffffffffffffffff");
});

test("bignumBelow2Pow64IsRejected", () => {
  assert.throws(() => encodeRfc8949(Bignum("positive", "5")), EncodeError);
  assert.throws(() => encodeRfc8949(Bignum("negative", "9223372036854775808")), EncodeError);
});

test("tagWrapsInnerValue", () => {
  assert.deepEqual(encodeRfc8949(Tag(100n, Int("5"))), new Uint8Array([0xd8, 0x64, 0x05]));
});

test("tagAtULongMaxDoesNotTruncate", () => {
  // Regression: verifies the tag argument doesn't get silently truncated/wrapped
  // when it's the maximum 64-bit unsigned value.
  assert.equal(hex(Tag(18446744073709551615n, Int("5"))), "dbffffffffffffffff05");
});

test("mapDuplicateKeysAreNotDeduped", () => {
  // rfc8949 has NO key-collision dedup (unlike dcbor) — both entries for a
  // literal duplicate key must survive to the output, in original order.
  assert.equal(hex(Map_([[Int("5"), Int("1")], [Int("5"), Int("2")]])), "a205010502");
});

File: adapters/go/rfc8949_test.go
package main

import (
	"encoding/hex"
	"testing"
)

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex %q: %v", s, err)
	}
	return b
}

func mustEncodeRfc8949(t *testing.T, v LogicalValue) []byte {
	t.Helper()
	b, err := encodeRfc8949(v)
	if err != nil {
		t.Fatalf("encodeRfc8949(%#v) error: %v", v, err)
	}
	return b
}

func assertHexEqual(t *testing.T, got []byte, wantHex string) {
	t.Helper()
	if hex.EncodeToString(got) != wantHex {
		t.Fatalf("got %x, want %s", got, wantHex)
	}
}

func TestRfc8949ShortestIntZero(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "0"}), "00")
}

func TestRfc8949ShortestIntBoundary23_24(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "23"}), "17")
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "24"}), "1818")
}

func TestRfc8949ShortestIntBoundary255_256(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "255"}), "18ff")
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "256"}), "190100")
}

func TestRfc8949ShortestIntBoundary65535_65536(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "65535"}), "19ffff")
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "65536"}), "1a00010000")
}

func TestRfc8949ShortestIntBoundary4294967295_4294967296(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "4294967295"}), "1affffffff")
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "4294967296"}), "1b0000000100000000")
}

func TestRfc8949NegativeInt(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "-1"}), "20")
}

func TestRfc8949NegativeIntMinI64Boundary(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, IntValue{Value: "-9223372036854775808"}), "3b7fffffffffffffff")
}

func TestRfc8949TextAndBytes(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, TextValue{Value: "a"}), "6161")
	assertHexEqual(t, mustEncodeRfc8949(t, BytesValue{Value: "deadbeef"}), "44deadbeef")
}

func TestRfc8949BoolAndNull(t *testing.T) {
	assertHexEqual(t, mustEncodeRfc8949(t, BoolValue{Value: true}), "f5")
	assertHexEqual(t, mustEncodeRfc8949(t, BoolValue{Value: false}), "f4")
	assertHexEqual(t, mustEncodeRfc8949(t, NullValue{}), "f6")
}

func TestRfc8949ArrayOfInts(t *testing.T) {
	arr := ArrayValue{Items: []LogicalValue{IntValue{Value: "1"}, IntValue{Value: "2"}}}
	assertHexEqual(t, mustEncodeRfc8949(t, arr), "820102")
}

func TestRfc8949NanCanonicalPayload(t *testing.T) {
	v := FloatValue{Width: "auto", Value: "NaN"}
	assertHexEqual(t, mustEncodeRfc8949(t, v), "f97e00")
}

func TestRfc8949NanCanonicalPayloadUnderExplicitWidthForcing(t *testing.T) {
	for _, width := range []string{"f16", "f32", "f64"} {
		v := FloatValue{Width: width, Value: "NaN"}
		assertHexEqual(t, mustEncodeRfc8949(t, v), "f97e00")
	}
}

func TestRfc8949NegativeZeroPreservedDistinctFromZero(t *testing.T) {
	neg := FloatValue{Width: "auto", Value: "-0.0"}
	pos := FloatValue{Width: "auto", Value: "0.0"}
	negBytes := mustEncodeRfc8949(t, neg)
	posBytes := mustEncodeRfc8949(t, pos)
	assertHexEqual(t, negBytes, "f98000")
	assertHexEqual(t, posBytes, "f90000")
	if hex.EncodeToString(negBytes) == hex.EncodeToString(posBytes) {
		t.Fatal("expected -0.0 and 0.0 to encode differently")
	}
}

func TestRfc8949ShortestFloatFormF16Exact(t *testing.T) {
	v := FloatValue{Width: "auto", Value: "2.5"}
	assertHexEqual(t, mustEncodeRfc8949(t, v), "f94100")
}

func TestRfc8949ExplicitFloatWidthForcesEncoding(t *testing.T) {
	v := FloatValue{Width: "f64", Value: "2.5"}
	assertHexEqual(t, mustEncodeRfc8949(t, v), "fb4004000000000000")
}

func TestRfc8949ExplicitFloatWidthRaisesOnPrecisionLoss(t *testing.T) {
	v := FloatValue{Width: "f16", Value: "0.1"}
	if _, err := encodeRfc8949(v); err == nil {
		t.Fatal("expected error")
	}
}

func TestRfc8949MapKeysSortedByEncodedBytes(t *testing.T) {
	value := MapValue{Entries: []MapEntry{
		{Key: IntValue{Value: "9"}, Val: IntValue{Value: "1"}},
		{Key: IntValue{Value: "10"}, Val: IntValue{Value: "2"}},
	}}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "a209010a02")
}

func TestRfc8949MapKeysPureBytewiseNoLengthPresort(t *testing.T) {
	value := MapValue{Entries: []MapEntry{
		{Key: IntValue{Value: "-24"}, Val: IntValue{Value: "1"}},
		{Key: IntValue{Value: "1000"}, Val: IntValue{Value: "2"}},
	}}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "a21903e8023701")
}

func TestRfc8949BignumAt2Pow64UsesTag2(t *testing.T) {
	value := BignumValue{Sign: "positive", Value: "18446744073709551616"}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "c249010000000000000000")
}

func TestRfc8949NegativeBignumOffsetByOne(t *testing.T) {
	value := BignumValue{Sign: "negative", Value: "18446744073709551616"}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "c348ffffffffffffffff")
}

func TestRfc8949BignumBelow2Pow64IsRejected(t *testing.T) {
	positive := BignumValue{Sign: "positive", Value: "5"}
	if _, err := encodeRfc8949(positive); err == nil {
		t.Fatal("expected error")
	}
	negative := BignumValue{Sign: "negative", Value: "9223372036854775808"}
	if _, err := encodeRfc8949(negative); err == nil {
		t.Fatal("expected error")
	}
}

func TestRfc8949TagWrapsInnerValue(t *testing.T) {
	value := TagValue{Tag: 100, Value: IntValue{Value: "5"}}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "d86405")
}

func TestRfc8949TagAtUint64MaxDoesNotTruncate(t *testing.T) {
	value := TagValue{Tag: 18446744073709551615, Value: IntValue{Value: "5"}}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "dbffffffffffffffff05")
}

func TestRfc8949MapDuplicateKeysAreNotDeduped(t *testing.T) {
	// rfc8949 (P4/P9) has no key-collision dedup logic, unlike dcbor -- both
	// entries for a literal duplicate key must survive to the output.
	value := MapValue{Entries: []MapEntry{
		{Key: IntValue{Value: "5"}, Val: IntValue{Value: "1"}},
		{Key: IntValue{Value: "5"}, Val: IntValue{Value: "2"}},
	}}
	assertHexEqual(t, mustEncodeRfc8949(t, value), "a205010502")
}
